src/Utilities/CssDocument.php

```diff
@@ -0,0 +1,283 @@
+     * be valid.
+     *
+     * @param string $atIdentifier name of the at-rule with the preceding at sign
+     * @param string $rule full content of the rule, including the at-identifier
+     *
+     * @return bool
+     */
+    private function isValidAtRule(string $atIdentifier, string $rule): bool
+    {
+        if (\strcasecmp($atIdentifier, '@font-face') === 0) {
+            return \stripos($rule, 'font-family') !== false && \stripos($rule, 'src') !== false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Splits input CSS code into an array of parts for different media queries, in order.
+     * Each part is an array where:
+     *
+     * - key "css" will contain clean CSS code (for @media rules this will be the group rule body within "{...}")
+     * - key "media" will contain "@media " followed by the media query list, for all allowed media queries,
+     *   or an empty string for CSS not within a media query
+     *
+     * Example:
+     *
+     * The CSS code
+     *
+     *   "@import "file.css"; h1 { color:red; } @media { h1 {}} @media tv { h1 {}}"
+     *
+     * will be parsed into the following array:
+     *
+     *   0 => [
+     *     "css" => "h1 { color:red; }",
+     *     "media" => ""
+     *   ],
+     *   1 => [
+     *     "css" => " h1 {}",
+     *     "media" => "@media "
+     *   ]
+     *
+     * @param array<array-key, string> $allowedMediaTypes
+     *
+     * @return array<int, array{css: string, media: string}>
+     */
+    private function splitCssAndMediaQuery(array $allowedMediaTypes): array
+    {
+        $mediaTypesExpression = '';
+        if (!empty($allowedMediaTypes)) {
+            $mediaTypesExpression = '|' . \implode('|', $allowedMediaTypes);
+        }
+
+        $cssSplitForAllowedMediaTypes = \preg_split(
+            '#(@media\\s++(?:only\\s++)?+(?:(?=[{(])' . $mediaTypesExpression . ')' . self::MEDIA_RULE_BODY_MATCHER
+            . ')#misU',
+            $this->styleRules,
+            -1,
+            PREG_SPLIT_DELIM_CAPTURE
+        );
+
+        // filter the CSS outside/between allowed @media rules
+        $cssCleaningMatchers = [
+            'import/charset directives' => '/\\s*+@(?:import|charset)\\s[^;]++;/i',
+            'remaining media enclosures' => '/\\s*+@media\\s' . self::MEDIA_RULE_BODY_MATCHER . '/isU',
+        ];
+
+        $splitCss = [];
+        foreach ($cssSplitForAllowedMediaTypes as $index => $cssPart) {
+            $isMediaRule = $index % 2 !== 0;
+            if ($isMediaRule) {
+                \preg_match('/^([^{]*+){(.*)}[^}]*+$/s', $cssPart, $matches);
+                $splitCss[] = [
+                    'css' => $matches[2],
+                    'media' => $matches[1],
+                ];
+            } else {
+                $cleanedCss = \trim(\preg_replace($cssCleaningMatchers, '', $cssPart));
+                if ($cleanedCss !== '') {
+                    $splitCss[] = [
```

Comment:
Let's compact this array to one line so the method becomes less long (vertically speaking).

Reply:
Done.